src/data/sequence.rs

```diff
@@ -0,0 +1,727 @@
+        self.owners.len() as u64
+    }
+
+    /// Return the expected access list version.
+    pub fn expected_access_list_version(&self) -> u64 {
+        self.access_list.len() as u64
+    }
+
+    pub fn versions(&self) -> ExpectedVersions {
+        ExpectedVersions::new(
+            self.expected_data_version(),
+            self.expected_owners_version(),
+            self.expected_access_list_version(),
+        )
+    }
+
+    /// Returns the data shell - that is - everything except the Values themselves.
+    pub fn shell(&self, expected_data_version: impl Into<Version>) -> Result<Self> {
+        let expected_data_version = to_absolute_version(
+            expected_data_version.into(),
+            self.expected_data_version() as usize,
+        )
+        .ok_or(Error::NoSuchEntry)? as u64;
+
+        let access_list = self
+            .access_list
+            .iter()
+            .filter(|a| a.expected_data_version() <= expected_data_version)
+            .cloned()
+            .collect();
+
+        let owners = self
+            .owners
+            .iter()
+            .filter(|owner| owner.expected_data_version <= expected_data_version)
+            .cloned()
+            .collect();
+
+        Ok(Self {
+            address: self.address,
+            data: Vec::new(),
+            access_list,
+            owners,
+            _flavour: self._flavour,
+        })
+    }
+
+    /// Return a value for the given Version (if it is present).
+    pub fn get(&self, version: Version) -> Option<&Value> {
+        let absolute_version = to_absolute_version(version, self.data.len())?;
+        self.data.get(absolute_version)
+    }
+
+    /// Return the current data entry (if it is present).
+    pub fn current_data_entry(&self) -> Option<DataEntry> {
+        match self.data.last() {
+            Some(value) => Some(DataEntry::new(self.data.len() as u64, value.to_vec())),
+            None => None,
+        }
+    }
+
+    /// Get a range of values within the given versions.
+    pub fn in_range(&self, start: Version, end: Version) -> Option<Values> {
+        let range = to_absolute_range(start, end, self.data.len())?;
+        Some(self.data[range].to_vec())
+    }
+
+    /// Return all Values.
+    pub fn values(&self) -> &Values {
+        &self.data
+    }
+
+    /// Get owner at version.
+    pub fn owner_at(&self, version: impl Into<Version>) -> Option<&Owner> {
+        let version = to_absolute_version(version.into(), self.owners.len())?;
+        self.owners.get(version)
+    }
+
+    /// Returns history of all owners
+    pub fn owner_history(&self) -> Vec<Owner> {
+        self.owners.clone()
+    }
+
+    /// Get history of owners within the range of versions specified.
+    pub fn owner_history_range(&self, start: Version, end: Version) -> Option<Vec<Owner>> {
+        let range = to_absolute_range(start, end, self.owners.len())?;
+        Some(self.owners[range].iter().copied().collect())
```

Comment:
Some comment about the copy.
Also is there a specific reason why this returns an `Option`
If the range doesn't exist then it could return an empty vector instead?

Reply:
I just checked `to_absolute_range` and it returns Option there so it's fine to return an option here. :+1:

Reply:
About copy: This was more or less the thing I was able to get through the compiler :)
It would seem that a reference to the slice would be most efficient, but I didn't manage to get that working. We could do a HO and look at it together to try figure it out?